### PR TITLE
B200-tuned factored LoRA MLP kernel behind lora_mlp_kernel_b200 + dedicated B200 CI suite

### PR DESCRIPTION
**Pull request by winglian**
# Description

Integrates the B200 (sm_100)-optimized factored LoRA MLP kernel from the B200 kernel-optimization campaign, behind a new opt-in `lora_mlp_kernel_b200` flag, plus a dedicated Modal CI lane that runs the sm_100-gated tests on an actual B200.

**Kernel** (`src/axolotl/kernels/blackwell/`): factored (no-merge) LoRA SwiGLU MLP — cuBLAS GEMMs + stride-aware fused SwiGLU Triton kernels, saved forward intermediates (no X/h re-reads in backward), deduplicated/concatenated skinny adapter GEMMs, 1-write+2-RMW dX accumulation, big-GEMM-first dispatch ordering, and scalar multiplies fused into addmm alpha. Certified on B200: **1.01–1.17x speed** vs the reference factored implementation at **~1.00–1.01x peak memory**, **29 vs 38 kernel launches** per fwd+bwd step (bf16, Llama-3-8B MLP shape, ranks 8–64, grad-accum K 1–8).

**Gating** (3 layers, opt-in only — never auto-enabled):
- config validation: requires `adapter: lora` + `lora_mlp_kernel`; rejects QLoRA/4-bit/8-bit, `lora_dropout > 0`, DoRA, FSDP/DeepSpeed; warns early when the detected capability isn't sm_100
- patch time: hard runtime gate on `torch.cuda.get_device_capability() == (10, 0)` with a logged fallback to the standard `lora_mlp_kernel` path on every other device; per-module checks (bf16 dense weights, adapters on all three projections, no biases, no quant)
- runtime: the wrapper degrades to the standard kernel if module state changes post-patch, and is `torch.compiler.disable`'d (Dynamo can't trace matmul into non-contiguous `out=` slices, so compiled models graph-break around the MLP instead of failing)

`CUBLAS_WORKSPACE_CONFIG=":4096:2"` (the certified cap) is applied best-effort at config validation and never clobbers an explicit value; the launch-environment recommendation is documented in `docs/lora_optims.qmd`.

**CI**: new `b200` pytest marker + `cicd/e2e_b200.py` Modal app (hardcodes `gpu="B200:1"` so a misconfigured GPU type can't silently skip-and-pass). The docker-e2e and nightly workflows get a cu130-only, single-matrix-entry B200 job running only `-m b200` tests under the certified launch env; The arch-portable correctness tests also run in the general GPU lanes (where the sm_100-only goldens skip), which keeps the real-device fallback test covered on non-B200 hardware.

**Not included**: the campaign's two merge-family alternate kernels (need `refresh_merge()` wired into the optimizer-step lifecycle; not certified head-to-head against the shipped kernel) — possible follow-up.

## Motivation and Context

LoRA fine-tuning on B200s currently uses the generic `lora_mlp_kernel` path. The handed-off campaign kernel is measurably faster at parity memory on that hardware, with an evidence trail (launch budgets, fp32 parity, perf certifications) worth locking in as CI goldens.

## How has this been tested?

- 62 new tests: fp32-reference parity at ranks 8/16/32/64 (< 2% Frobenius vs an fp32 reference of the same factored form), live-gradients canary, addmm-alpha scalar-fusion guard, equivalence vs the standard `LoRA_MLP` kernel, patch-selection/fallback tests, CUBLAS env-var contract subprocess tests, config-validation tests, and an sm_100-keyed launch-count/kernel-name budget (`tests/e2e/kernels/blackwell_launch_budgets.json`, golden-update procedure in `_meta`)
- Verified on sm_120 hardware (RTX PRO 6000 Blackwell): all arch-portable correctness tests pass, and the arch gate falls back through the real patch path (12 passed, 2 sm_100-only skips)
- Existing suites unaffected: 36 lora-kernel-patching tests, 220 CPU kernel tests, 74 validation tests all pass; pre-commit clean
- First B200 CI run will exercise the launch-budget golden on real sm_100; the perf-vs-standard-kernel gate skips and reports its measured ratio until a threshold is certified into the budgets JSON (deliberately not fabricated)

Still pending before any default-on consideration (per the handoff's mandatory gate): a loss-curve equivalence run on a real B200 (same seed/data/optimizer, standard vs B200 kernel, few hundred steps).

## AI Usage Disclaimer

Yes — Claude Code

## Types of changes

- New feature (opt-in kernel path + config flag)
- CI (new B200 Modal suite, nightly lane)
- Docs (`docs/lora_optims.qmd`)

### CODE REVIEW

File: src/axolotl/kernels/blackwell/lora_mlp_factored.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        d_gate_A = torch.matmul(dXAg.t(), X)  # [r, HIDDEN]
+
+        # dX: 1 write + 2 RMW passes. When inplace, dX reuses X's storage --
+        # safe because every read of X (d_up_A/d_gate_A above) has already
+        # happened, and it saves a full [T, HIDDEN] allocation.
+        dX = X if ctx.inplace else torch.empty_like(X)
+        torch.matmul(grad_up, up_W, out=dX)
+        dX.addmm_(grad_gate, gate_W)
+        dXA_gu = torch.cat([dXAg, dXAu], dim=1)  # [T, 2r]
+        dX.addmm_(dXA_gu, A_gu)
```
[the remark this answers is not in the document]
Comment (winglian) [reply src/axolotl/kernels/blackwell/lora_mlp_factored.py:177]: Valid — the Triton in-place writes to `buf_eg` bypass autograd's version counters, so a `retain_graph` re-backward would silently reuse mutated buffers. Added the one-shot guard at the top of `backward` (before unpacking) plus a regression test (`test_second_backward_through_same_graph_raises`) in c9c5ef856.

File: src/axolotl/utils/schemas/config.py [hunk 2]
```diff
@@ -1871,6 +1877,65 @@ def check_multigpu_lora_kernels(cls, data):
...
+
+        # Best-effort: cuBLAS reads this env var at handle creation, so the
+        # launch environment is the reliable place to set it (documented in
+        # docs/lora_optims.qmd); this covers the common single-process case.
+        from axolotl.kernels.blackwell.support import (
+            maybe_set_cublas_workspace_config,
+        )
+
+        maybe_set_cublas_workspace_config()
+        return data
```
[the remark this answers is not in the document]
Comment (winglian) [reply src/axolotl/utils/schemas/config.py:1946]: Agreed — setting a process-wide cuBLAS workspace cap on hardware that will never dispatch this kernel is an unwanted global side effect. The best-effort call is now gated on `compute_capability` being sm_100 or unknown (unknown keeps the certified path for validation contexts without GPU info), with tests for both branches. Fixed in c9c5ef856.

File: tests/e2e/kernels/test_blackwell_lora_mlp.py [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+        }
+    )
+    model = _tiny_peft_llama()
+    apply_lora_kernel_patches(model, cfg)
+    return get_layers(model)[0].mlp.forward
+
+
+@pytest.mark.skipif(
+    torch.cuda.get_device_capability() == (10, 0), reason="requires a non-B200 device"
+)
```
[the remark this answers is not in the document]
Comment (winglian) [reply tests/e2e/kernels/test_blackwell_lora_mlp.py]: Correct — skipif args evaluate at collection and `get_device_capability()` raises on CPU-only hosts. Moved the check into the test body (the module-level CUDA skipif still guards execution); verified `pytest --collect-only` with `CUDA_VISIBLE_DEVICES=""` now collects cleanly. Fixed in c9c5ef856.

**Comment by winglian** [comment]
Addressed the CodeRabbit review in c9c5ef856. The three actionable comments were all valid and are fixed (replied inline). Dispositions on the seven nitpicks:

**Taken:**
- **bf16 config-time warning** (`config.py`): now warns upfront when neither `bf16` nor `bfloat16` is set, mirroring the compute-capability warning.
- **Marker scope vs docstring** (`test_blackwell_lora_mlp.py`): real coverage gap — beyond the docstring mismatch, `test_patching_falls_back_on_non_sm100` skips on B200 and was deselected everywhere else, so the real-device fallback path had no CI lane at all. Fixed by dropping the `-m "not b200"` deselect from the general e2e lane (cicd.sh reverted to its original block): the arch-portable tests + fallback test now run on L40S, while the module-level `b200` mark still scopes the dedicated B200 lane via `-m b200`. Docstring updated to say exactly this.
- **Hoist `lora_mlp_b200_module_supported` import** out of the per-layer loop (`lora_kernels.py`).
- **SwiGLU view-kernel stride contract** (`swiglu_gateup.py`): added an explicit same-shape/unit-last-stride check raising `ValueError` — cheap insurance for future callers of an otherwise silently-corrupting contract.

**Declined:**
- **`always()` on `docker-e2e-cleanup`**: a bare `always()` would make cleanup spin a Modal container even when `gate-skip-e2e` skipped the whole matrix (unlabeled PRs), a bigger behavior change than the one being fixed; and the job is best-effort pruning of >1-day-old HF-cache files that any later green run performs anyway. Happy to revisit cleanup gating across all three lanes in a follow-up.
- **RUF059 underscore-prefixing** in `backward`: the RUF rule set isn't enabled in this repo's ruff config, and the explicit names document the save-list layout 1:1 with `forward`.
- **`save_for_backward` for base weights**: intentionally matches the existing `LoRA_MLP` pattern in `kernels/lora.py` (`ctx.weights = (...)`). The bases are frozen in this kernel's supported configs (version tracking is moot), and routing large frozen weights through saved-tensor hooks would make activation-offload hooks treat them as offloadable activations.
